
<launch>

  <!-- ROBOT LOCALIZATION VERSION: use this with ROS bag demo_mapping.bag -->
  <!-- A database "~/.ros/rtabmap.db" must be already created from -->
  <!-- the "demo_robot_mapping.launch" demo -->
  <!-- Once RTAB-Map GUI started, you can do "Edit->Download Map" to get all the map in the GUI -->
   
  <!-- Choose visualization -->
  <arg name="database_path"     default="rtabmap.db"/>
  <arg name="rviz" default="false" />
  <arg name="rtabmapviz" default="false" /> 
  <arg name="wait_for_transform_duration"  default="0.5"/>
  <arg name="localization"        default="true"/>
  
  <param name="use_sim_time" type="bool" value="false"/>

  <group ns="rtabmap">
    <!-- SLAM (robot side) -->
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="">
	  <param name="frame_id" type="string" value="base_footprint"/>
	  <param name="odom_frame_id" type="string" value="odom"/>

	  <param name="wait_for_transform_duration"  type="double"   value="$(arg wait_for_transform_duration)"/>
	  
	  <param name="cloud_ceiling_culling_height"  type="double"   value="2.5"/>
	
	  <param name="subscribe_depth" type="bool" value="true"/>
	  <param name="subscribe_laserScan" type="bool" value="true"/>
	
	  <remap from="odom" to="/odom_throttled"/>
	  <remap from="scan" to="/scan_throttled"/>

	  <remap from="rgb/image" to="/data_throttled_image"/>
  	  <remap from="depth/image" to="/data_throttled_image_depth"/>
  	  <remap from="rgb/camera_info" to="/data_throttled_camera_info"/>
  	
      <param name="rgb/image_transport" type="string" value="compressed"/>
      <param name="depth/image_transport" type="string" value="compressedDepth"/>
	
	  <!-- RTAB-Map's parameters: do "rosrun rtabmap rtabmap (double-dash)params" to see the list of available parameters. -->
	  <param name="database_path"       type="string" value="$(arg database_path)"/>  <!-- Database used for localization -->
      <param name="Rtabmap/DetectionRate" type="string" value="1"/>     <!-- Don't need to do relocation very often! Though better results if the same rate as when mapping. -->
      <param name="Mem/STMSize" type="string" value="1"/>               <!-- 1 location in short-term memory -->
      <param name="RGBD/PoseScanMatching" type="string" value="true"/>  <!-- Do odometry correction with consecutive laser scans -->
	  <param name="RGBD/LocalLoopDetectionSpace" type="string" value="true"/>
	  <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/> 
	  <param name="LccIcp/Type" type="string" value="2"/>                      <!-- 0=No ICP, 1=ICP 3D, 2=ICP 2D -->
	  <param name="LccIcp2/MaxFitness" type="string" value="10"/>
	  <param name="LccReextract/MaxDepth" type="string" value="0.0"/>          <!-- 3D visual words maximum depth 0=infinity -->
	  <param name="LccBow/InlierDistance" type="string" value="0.1"/>          <!-- 3D visual words correspondence distance -->
      <param name="LccIcp2/CorrespondenceRatio" type="string" value="0.2"/>

	  <!-- localization mode -->
	  <param     if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
	  <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>
	  <param name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)"/>  
    </node>
  
    <!-- Visualisation RTAB-Map -->
    <node if="$(arg rtabmapviz)" pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
  	  <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_laserScan" type="bool" value="true"/>
      <param name="frame_id" type="string" value="base_footprint"/>
	  <param name="wait_for_transform_duration"  type="double"   value="$(arg wait_for_transform_duration)"/>
    
      <remap from="rgb/image" to="/data_throttled_image"/>
      <remap from="depth/image" to="/data_throttled_image_depth"/>
      <remap from="rgb/camera_info" to="/data_throttled_camera_info"/>
      <remap from="scan" to="/scan_throttled"/>
      <remap from="odom" to="/odom_throttled"/>
      
      <param name="rgb/image_transport" type="string" value="compressed"/>
      <param name="depth/image_transport" type="string" value="compressedDepth"/>
    </node>
  </group>

  <!-- Visualisation RVIZ -->
  <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find piko_nav)/config/rtabmap.rviz" output="screen"/>
   <node pkg="nodelet" type="nodelet" name="points_xyzrgb" args="standalone rtabmap_ros/point_cloud_xyzrgb">
    <remap from="rgb/image"       to="/data_throttled_image"/>
    <remap from="depth/image"     to="/data_throttled_image_depth"/>
    <remap from="rgb/camera_info" to="/data_throttled_camera_info"/>
    <remap from="cloud"           to="voxel_cloud" />
    
    <remap from="odom" to="/odom_throttled"/>
    <remap from="scan" to="/scan_throttled"/>

    <param name="rgb/image_transport" type="string" value="compressed"/>
    <param name="depth/image_transport" type="string" value="compressedDepth"/>
    
    <param name="queue_size" type="int" value="10"/>
    <param name="voxel_size" type="double" value="0.01"/>
  </node>

</launch>
